TombEngine/Objects/TR2/Entity/tr2_sword_guardian.cpp

```diff
@@ -12,210 +13,372 @@
 			(GetRandomControl() * 256 / 32768) + item->Pose.Position.x - 128,
 			(GetRandomControl() * 256 / 32768) + item->Pose.Position.y - 256,
 			(GetRandomControl() * 256 / 32768) + item->Pose.Position.z - 128);
 
-		TriggerGunSmoke(pos.x, pos.y, pos.z, 1, 1, 1, 1, LaraWeaponType::GrenadeLauncher, 32);
+		SpawnSpearGuardianSmoke(pos, item->RoomNumber);
 		SoundEffect(SFX_TR2_WARRIOR_HOVER, &item->Pose);
 	}
 
+	void InitialiseSwordGuardian(short itemNumber)
+	{
+		auto* item = &g_Level.Items[itemNumber];
+		InitialiseCreature(itemNumber);
+		item->ItemFlags[0] = 0; // Joint index when swapping mesh.
+		item->ItemFlags[1] = 1; // Immune state. true = immune to damage.
+		item->ItemFlags[2] = 1; // If 1, swap to jade. If 2, swap to normal.
+		item->ItemFlags[3] = 0; // If mesh is swapped to jade, then it's true. Otherwise false.
+		SetAnimation(item, SWORD_GUARDIAN_ANIM_AWAKE);
+		item->Status &= ~ITEM_INVISIBLE; // This statue is draw by default.
+
+		const auto& object = Objects[item->ObjectNumber];
+		for (int jointIndex = 0; jointIndex < object.nmeshes; jointIndex++)
+			SwapMeshToJade(*item, jointIndex, false);
+	}
+
 	void SwordGuardianControl(short itemNumber)
 	{
 		if (!CreatureActive(itemNumber))
 			return;
 
 		auto* item = &g_Level.Items[itemNumber];
+		auto* object = &Objects[item->ObjectNumber];
 		auto* creature = GetCreatureInfo(item);
 
 		short angle = 0;
-		short head = 0;
+		short neck = 0;
 		short torso = 0;
 
-		bool isLaraAlive = LaraItem->HitPoints > 0;
+		bool isLaraAlive = creature->Enemy != nullptr && creature->Enemy->IsLara() && creature->Enemy->HitPoints > 0;
 
 		if (item->HitPoints <= 0)
 		{
-			if (item->Animation.ActiveState != 12)
+			if (item->Animation.ActiveState != SWORD_GUARDIAN_STATE_DEATH)
 			{
-				SoundEffect(SFX_TR4_EXPLOSION1, &LaraItem->Pose);
-				SoundEffect(SFX_TR4_EXPLOSION2, &LaraItem->Pose);
-				ExplodingDeath(itemNumber, BODY_EXPLODE);
-				DisableEntityAI(itemNumber);
-				KillItem(itemNumber);
-				item->Animation.ActiveState = 12;
+				item->ItemFlags[3] = 1;
+				item->ItemFlags[0] = object->nmeshes - 1;
+				item->Animation.ActiveState = SWORD_GUARDIAN_STATE_DEATH;
 			}
 
+			if (DoSwordGuardianSwapMesh(*item))
+				CreatureDie(itemNumber, true);
 			return;
 		}
 		else
 		{
+
+			AI_INFO AI;
 			creature->LOT.Step = CLICK(1);
 			creature->LOT.Drop = -CLICK(1);
 			creature->LOT.Fly = NO_FLYING;
 			creature->LOT.Zone = ZoneType::Basic;
-
-			AI_INFO AI;
 			CreatureAIInfo(item, &AI);
-
-			if (item->Animation.ActiveState == 8)
+			if (AI.enemyZone != AI.zoneNumber && item->Animation.ActiveState == SWORD_GUARDIAN_STATE_FLY)
 			{
-				if (AI.zoneNumber != AI.enemyZone)
-				{
-					creature->LOT.Step = SECTOR(20);
-					creature->LOT.Drop = -SECTOR(20);
-					creature->LOT.Fly = CLICK(0.25f);
-					creature->LOT.Zone = ZoneType::Flyer;
-					CreatureAIInfo(item, &AI);
-				}
+				creature->LOT.Step = BLOCK(20);
+				creature->LOT.Drop = -BLOCK(20);
+				creature->LOT.Fly = 64;
```

Comment:
why was the fly changed from CLICK(0.25f) to 64?

Reply:
Fly is not a height but a speed, and 64 is just `1 click / 4`